### PR TITLE
Improve button behavior when multiple mouse buttons are used at the same time

### LINKED ISSUE #93395: Button can still be clicked if overlapped with another control with MOUSE_FILTER_STOP if the user does not release mouse button(-s)
### Tested versions

- Reproducible in `v4.2.2.stable.official [15073afe3]`

### System information

Godot v4.2.2.stable - Windows 10.0.19045 - Vulkan (Forward+) - dedicated NVIDIA GeForce GTX 1660 (NVIDIA; 31.0.15.3623) - AMD Ryzen 7 2700X Eight-Core Processor (16 Threads)

### Issue description

I have a number of GUI-s. When I click on a button, it opens a fullscreen blocking GUI (mouse_filter = STOP). However, it seems to be still possible to press this initial button, **even if it is completely blocked by the GUI** (any control with MOUSE_FILTER_STOP).

A workaround would be to always check `is_hovered` first, however, the button UI will display presses anyway. If you want to NOT display clicks at all, you will need your own button...

I'm not sure about the severity of the issue, but in some projects this may mean that Godot's buttons should never be used at all (or disable styles, textures, etc. to never let it display clicks), and instead you should always use your own buttons.

### Steps to reproduce

1. Launch project
2. Press the `Click me` button with RIGHT or MIDDLE mouse button. Hold it (don't release!)
3. Now click the button with LEFT mouse button. Keep the RIGHT / MIDDLE button pressed
4. The panel pops up and blocks the button. Keep the RIGHT / MIDDLE button pressed
5. While still holding the RIGHT / MIDDLE button down, start clicking LEFT mouse button multiple times.
6. **Panel lets the clicks come through** to the button. This should NOT happen.

p.6 won't happen if you don't hold RIGHT / MIDDLE button at p.2
If you try the steps above with `check_is_hovered = true` (property on `click_me_button.gd`), you will see that the behavior is mainly valid (except that button UI still displays clicks)

### Minimal reproduction project (MRP)

[button_mrp.zip](https://github.com/user-attachments/files/15916425/button_mrp.zip)

### PR DESCRIPTION
**Pull request by anniryynanen**
- To emit `pressed`, buttons require that the press was initiated while hovering.
- Controls can't grab focus from a mouse click if they're not hovered.
- Hovers are updated both before and after a handled mouse button event.

Fixes #93395.

An updated MRP that makes testing this even easier: [button_mrp.zip](https://github.com/user-attachments/files/15942556/button_mrp.zip)

### CODE REVIEW

**Comment by anniryynanen** [comment]
Achievement unlocked: First unit test broken ^^;;

Fixing.

**Comment by anniryynanen** [comment]
I fixed the test failing in Windows, but I can't reproduce the linux Linux failure on `Arch Linux #1 SMP PREEMPT_DYNAMIC Sun, 16 Jun 2024 19:06:37 +0000 - Tty - GLES3 (Compatibility) - Mesa Intel(R) Graphics (ADL GT2) - 12th Gen Intel(R) Core(TM) i5-1235U (12 Threads)`.

I wonder if I could somehow make the segfault happen on my machine... Maybe a VM similar to the cloud build or something? Is there a feasible way to do that?

File: scene/gui/base_button.cpp [hunk 1]
```diff
@@ -144,7 +144,7 @@ void BaseButton::_toggled(bool p_pressed) {
 }
 
 void BaseButton::on_action_event(Ref<InputEvent> p_event) {
-	if (p_event->is_pressed()) {
+	if (p_event->is_pressed() && status.hovering) {
```
Comment (KoBeWi) [inline scene/gui/base_button.cpp:147]: This makes it impossible to press buttons with keyboard.

Comment (anniryynanen) [reply scene/gui/base_button.cpp:147]: Oo! I'll try to fix that.

**Review by Calinou — APPROVED** [review]
Tested locally, it works as expected.

### Before

*Notice how the hover stylebox lingers until I move the mouse again, even though the mouse is being blocked by the translucent control.*

https://github.com/godotengine/godot/assets/180032/dd09c969-1334-4b16-9384-0770530835c1

### After

https://github.com/godotengine/godot/assets/180032/5ae3e6e4-7e0d-4e93-ab25-4abcc19bd3aa

File: scene/gui/base_button.cpp [hunk 2]
```diff
@@ -144,7 +144,7 @@ void BaseButton::_toggled(bool p_pressed) {
 }
 
 void BaseButton::on_action_event(Ref<InputEvent> p_event) {
-	if (p_event->is_pressed()) {
+	if (p_event->is_pressed() && (!p_event->is_class("InputEventMouseButton") || status.hovering)) {
```
Comment (akien-mga) [inline scene/gui/base_button.cpp:147]: A cast should be more performant than string comparisons, so I would suggest moving `Ref<InputEventMouseButton> mouse_button = p_event;` from below at the top of the method, and use that.

**Comment by akien-mga** [comment]
I'm keeping this for 4.4 as changes to BaseButton and Viewport are historically prone to regressions, as this impacts a lot of the UI, both for the Godot editor and for games. We definitely need more unit and integration testing there to be able to make such changes with confidence.

So the change is probably good, but it can use a full dev cycle to confirm that, hence 4.4.

**Comment by anniryynanen** [comment]
Thanks for the reviews! I'll look at this next week, when I'm no longer consumed by Godot Wild Jam.

File: scene/gui/base_button.cpp [hunk 3]
```diff
@@ -174,7 +176,6 @@ void BaseButton::on_action_event(Ref<InputEvent> p_event) {
 	}
 
 	if (!p_event->is_pressed()) {
-		Ref<InputEventMouseButton> mouse_button = p_event;
 		if (mouse_button.is_valid()) {
 			if (!has_point(mouse_button->get_position())) {
 				status.hovering = false;
```
[the remark this answers is not in the document]
Comment (anniryynanen) [reply scene/gui/base_button.cpp]: I removed it, it seems ok. Or at least I didn't find any problems.

File: scene/main/viewport.cpp [hunk 4]
```diff
@@ -3256,6 +3261,12 @@ void Viewport::push_input(const Ref<InputEvent> &p_event, bool p_local_coords) {
 		_push_unhandled_input_internal(ev);
 	}
 
+	// Update mouse over again in case button event triggered changes to controls.
+	Ref<InputEventMouseButton> mb = ev;
+	if (mb.is_valid() && is_input_handled()) {
+		_update_mouse_over();
+	}
+
```
[the remark this answers is not in the document]
Comment (anniryynanen) [reply scene/main/viewport.cpp]: I'll leave it for another PR.

File: scene/gui/base_button.cpp [hunk 5]
```diff
@@ -174,12 +176,6 @@ void BaseButton::on_action_event(Ref<InputEvent> p_event) {
 	}
 
 	if (!p_event->is_pressed()) {
-		Ref<InputEventMouseButton> mouse_button = p_event;
-		if (mouse_button.is_valid()) {
-			if (!has_point(mouse_button->get_position())) {
-				status.hovering = false;
-			}
-		}
```
Comment (Sauermann) [inline scene/gui/base_button.cpp:182]: I don't see, how this removal has an impact on the MRPs of the initial issue description and of this PR.

Also wouldn't this removal create for certain projects situations, where the hovering status is kept `true`, while it should be `false`?

Perhaps this change could be investigated in a separate PR. The other two modifications are sufficient to resolve this problem in the issue description.

**Review by Sauermann — APPROVED** [review]
Code looks good.
Have tested with both available MRPs and didn't find any obvious problems.

File: scene/gui/base_button.cpp (hunk 5 above)
Comment (kitbdev) [reply scene/gui/base_button.cpp]: See https://github.com/godotengine/godot/pull/93500#discussion_r1713121040
This is because when clicking on a child with mouse filter pass, this can make the hovering state incorrect and then the button can no longer be clicked the second time.
So I think this should still be removed for this PR.

- I have also tested to make sure that https://github.com/godotengine/godot/issues/33553 doesn't reappear with this removed.

The appearance when pressed and hovering over child is also incorrect, but this can be looked into later.

Comment (anniryynanen) [reply scene/gui/base_button.cpp]: I had forgotten about the MOUSE_FILTER_PASS case, I'll re-remove this.

Comment (Sauermann) [reply scene/gui/base_button.cpp:182]: Thanks for the explanation. I agree.

**Comment by akien-mga** [comment]
Thanks!
